### ISSUE TITLE
[RFE] Common Linux System Roles Travis CI Setup

### ISSUE THREAD

**Issue by i386x** [linux-system-roles/template#4]
# Common Linux System Roles Travis CI Setup

**The following is just a draft and may be changed based on the comments below. The used code snippets may contain flaws and some constructs were not yet tested.**

This RFE will propose how to achieve to have the same [Travis CI](https://travis-ci.org/) setup across all [Linux system roles](https://github.com/linux-system-roles).

## Current Status

In [Linux system roles project](https://github.com/linux-system-roles), there are typically two approaches of how [Travis CI](https://travis-ci.org/) is set up (i.e. what is invoked by Travis depending on the content of `.travis.yml`):

- Travis runs the [Molecule](https://molecule.readthedocs.io/en/stable/). This is common for [`timesync`](https://github.com/linux-system-roles/timesync), [`kdump`](https://github.com/linux-system-roles/kdump), and [`selinux`](https://github.com/linux-system-roles/selinux) roles. As there is no Python code in these roles, there is no need to run Python linters and then [`.travis.yml`](https://github.com/linux-system-roles/kdump/blob/master/.travis.yml) just contains [Molecule](https://molecule.readthedocs.io/en/stable/) commands together with necessary environment settings to run them.

- Travis runs the [`tox`](https://tox.readthedocs.io/en/latest/index.html), from which are run Python linters and [Molecule](https://molecule.readthedocs.io/en/stable/). This approach is used by [`network`](https://github.com/linux-system-roles/network) role. Also, [`.travis.yml`](https://github.com/linux-system-roles/network/blob/master/.travis.yml) is a bit more complex than in the previous roles. It uses build matrix and `before_install` section. When a new pull request is made, Travis creates for every element of the build matrix a job with environment based on the element of the build matrix definition. Then it prepares environment by running the commands from `before_install` and `install` sections, respectivelly, and finally it fires up commands specified in `script`. In `script` there is just one command: `tox`. [`tox`](https://tox.readthedocs.io/en/latest/index.html) is a tool that looks for `tox.ini` and based on its content it runs specified commands, mostly Python linters, but as can be seen in [`network`'s `tox.ini`](https://github.com/linux-system-roles/network/blob/master/tox.ini), it can also run Molecule. Using [`tox-travis`](https://tox-travis.readthedocs.io/en/stable/) plugin, we can bind concrette test environments with concrette Travis jobs together, depending on the version of Python interpreter used.

The [`molecule` setup](https://github.com/linux-system-roles/template/tree/master/molecule/default) is common for all Linux system roles.

## Proposal

The proposed Travis CI setup that should be common for all Linux system roles is derived from `network` role.

The proposed general format of `.travis.yml` will be:

```yaml
# Ubuntu distro name (here xenial):
dist: xenial
language: python
matrix:
  include:
    - python: 2.6
      # For Python 2.6, use older distro (here trusty):
      dist: trusty
      # We use LSRENV (the name of the variable can be changed if desired) to
      # save the unique value (here the desired Python interpreter version) to
      # distinguish between Travis jobs. This can be usefull if some role wish
      # to perform a specific action depending on the job environment:
      env: LSRENV=2.6
      # For now, only network role use Python linters, so save Travis CI
      # resources if we just need to run Molecule. For Travis conditions
      # syntax, see https://docs.travis-ci.com/user/conditions-v1
      if: repo = "linux-system-roles/network"
    - python: 2.7
      env: LSRENV=2.7
      if: repo = "linux-system-roles/network"
    - python: 3.5
      env: LSRENV=3.5
      # Under 3.5, we run Molecule
    - python: 3.6
      env: LSRENV=3.6
      if: repo = "linux-system-roles/network"
    - python: 3.7
      env: LSRENV=3.7
      if: repo = "linux-system-roles/network"
    - python: 3.7-dev
      env: LSRENV=3.7-dev
      if: repo = "linux-system-roles/network"
    - python: 3.8-dev
      env: LSRENV=3.8-dev
      if: repo = "linux-system-roles/network"
    #- python: nightly
    #  env: LSRENV=nightly
    #  if: repo = "linux-system-roles/network"

services:
  - docker

before_install:
  # In .travis/before_install.sh can be commands that install packages specific
  # for the concrette role, or set specific environment variables. To decide
  # what to set or install in which environment, the value of LSRENV should be
  # helpfull:
  - if [ -f .travis/before_install.sh ]; then .travis/before_install.sh; fi

install:
  - pip install molecule docker tox tox-travis

script:
  - tox
```

The proposed general format of `tox.ini` will be:

```ini
[tox]
# In `network` role's `tox.ini`, there is the `envlist` variable holding the
# list of test environments whose tests are run when `tox` is launched. But
# when using tox-travis plugin, the test environments are listed in `[travis]`
# section, so I am not sure if `envlist` is necessary:
#envlist = ...
skipsdist = true
skip_missing_interpreters = true

[testenv]
basepython = python3
# List common dependencies for Python interpreters here:
deps =
    py{26,27,36,37,38}: pytest-cov
    py{27,36,37,38}: pytest>=3.5.1
    py{26,27}: mock
    py26: pytest

[base]
passenv = *
setenv =
    PYTHONPATH = {toxinidir}/library:{toxinidir}/module_utils
    LC_ALL = C
changedir = {toxinidir}/tests
covtarget = {toxinidir}/library --cov {toxinidir}/module_utils
pytesttarget = .

[testenv:black]
deps =
    black
commands =
    black --check --diff --include "^[^.].*\.py$" --exclude "/(\.[^.].*|tests/roles)/" .

[testenv:py26]
install_command =
    pip install {opts} {packages}
list_dependencies_command =
    pip freeze
basepython = python2.6
passenv = {[base]passenv}
setenv =
    {[base]setenv}
changedir = {[base]changedir}
commands =
    pytest --durations=5 \
           --cov={[base]covtarget} \
           --cov-report=html:htmlcov-py26 \
           --cov-report=term \
           {posargs} \
           {[base]pytesttarget}

[testenv:py27]
basepython = python2.7
passenv = {[base]passenv}
setenv =
    {[base]setenv}
changedir = {[base]changedir}
commands =
    pytest --durations=5 \
           --cov={[base]covtarget} \
           --cov-report=html:htmlcov-py27 \
           --cov-report=term \
           {posargs} \
           {[base]pytesttarget}

[testenv:py36]
basepython = python3.6
passenv = {[base]passenv}
setenv =
    {[base]setenv}
changedir = {[base]changedir}
commands =
    pytest --durations=5 \
           --cov={[base]covtarget} \
           --cov-report=html:htmlcov-py36 \
           --cov-report=term \
           {posargs} \
           {[base]pytesttarget}

[testenv:py37]
basepython = python3.7
passenv = {[base]passenv}
setenv =
    {[base]setenv}
changedir = {[base]changedir}
commands =
    pytest --durations=5 \
           --cov={[base]covtarget} \
           --cov-report=html:htmlcov-py37
           --cov-report=term \
           {posargs} \
           {[base]pytesttarget}

[testenv:py38]
basepython = python3.8
passenv = {[base]passenv}
setenv =
    {[base]setenv}
changedir = {[base]changedir}
commands =
    pytest --durations=5 \
           --cov={[base]covtarget} \
           --cov-report=html:htmlcov-py38 \
           --cov-report=term \
           {posargs} \
           {[base]pytesttarget}

# Custom role settings may set LSR_PYLINT_DIRS environment variable if there is
# a reason to have other than default directories (library/ module_utils/ and
# tests/). For example, network role should have this set like
#
#   LSR_PYLINT_DIRS="library/network_connections.py module_utils/network_lsr tests/unit/test_network_connections.py"
#
[testenv:pylint]
basepython = python2.7
setenv =
    {[base]setenv}
deps =
    pylint>=1.8.4
    ansible
commands =
    pylint --errors-only \
           {posargs} \
           {env:LSR_PYLINT_DIRS:library module_utils tests}

[testenv:flake8]
basepython = python2.7
deps =
    flake8>=3.5
whitelist_externals = flake8
commands =
    flake8 --statistics {posargs} .

[testenv:coveralls]
basepython = python2.7
passenv = TRAVIS TRAVIS_*
deps =
    coveralls
changedir = {[base]changedir}
commands =
    coveralls

# Here we provide a way how a role can add its custom command to be run. Such
# extra command is run at the end of each testenv run and is driven by
# environment variables. Involved environment variables are:
#
#   LSR_TEXTRA_DEPS
#     - contains dependency needed by commands to run smoothly; if more than
#       one dependency is needed, use external file together with '-r' option
#       (see PEP 508)
#
#   LSR_TEXTRA_DIR
#     - directory to which to cd
#
#   LSR_TEXTRA_CMD
#     - custom command to be run
#
# Example: `network` system role need to run `./tests/ensure_non_running_provider.py`
#          to check for the existence of `*_provider.yml` playbooks. The script
#          is run in Python 3.6.
#
#          To make this possible, we add to `.travis/before_install.sh` a snippet:
#
#              if [ x${LSRENV} = x3.6 ]; then
#                LSR_TEXTRA_DEPS='PyYAML'
#                LSR_TEXTRA_DIR='tests'
#                LSR_TEXTRA_CMD='./ensure_non_running_provider.py'
#              fi
#
[testenv:extra]
passenv = *
deps =
    {env:LSR_TEXTRA_DEPS}
changedir = {toxinidir}/{env:LSR_TEXTRA_DIR:.}
commands =
    {env:LSR_TEXTRA_CMD:python --version}

# LSR_MOLECULE_DEPS may contain aditional Molecule dependencies. For example,
# in `network` system role, LSR_MOLECULE_DEPS can be set as
#
#   LSR_MOLECULE_DEPS='-rmolecule_requirements.txt'
#
# where `molecule_requirements.txt` contains two lines:
#
#   jmespath
#   git+https://github.com/tyll/selinux-pypi-shim@fulllocation
#
[molecule]
deps =
    docker
    molecule
    {env:LSR_MOLECULE_DEPS}

[testenv:molecule_lint]
deps =
    {[molecule]deps}
commands_pre =
    molecule --version
    ansible --version
commands =
    molecule {posargs} lint

[testenv:molecule_syntax]
deps =
    {[molecule]deps}
commands =
    molecule {posargs} syntax

[testenv:molecule_test]
deps =
    {[molecule]deps}
commands =
    molecule {posargs} test

[pytest]
addopts = -rxs

[flake8]
show_source = true
max-line-length = 88
ignore = E402,W503

[pylint]
max-line-length = 88
disable = wrong-import-position

[pycodestyle]
max-line-length = 88

[travis]
python =
  2.6: py26,extra
  2.7: py27,coveralls,flake8,pylint,extra
  3.5: molecule_lint,molecule_syntax,molecule_test,extra
  3.6: py36,black,extra
  3.7: py37,extra
  3.8: py38,extra
```

Other configuration files used by Python linters can be reused from [`network` role](https://github.com/linux-system-roles/network).

**Comment by pcahyna** [comment]
>       # For now, only network role use Python linters, so save Travis CI
>       # resources if we just need to run Molecule. For Travis conditions
>       # syntax, see https://docs.travis-ci.com/user/conditions-v1
>       if: repo = "linux-system-roles/network"

Is there a way to generalize this condition? The storage role has Python modules as well and could benefit from Python linters (and the selinux role too).

**Comment by pcahyna** [comment]
@vcrhonek @paulfantom what do you think?

**Comment by i386x** [comment]
Maybe
```yaml
...
env:
  global:
    LSR_HASPYTHON="/^linux-system-roles\/(network|storage|selinux)$/"
matrix:
  include:
    - python: 2.6
      ...
      if: repo =~ env(LSR_HASPYTHON)
      ...
...
```
could help to generalize the condition.

**Comment by pcahyna** [comment]
Could we put something in the role itself to mark it as "using Python"? Would be enough to just test for the presence of `{library,module_utils,*plugins}/*.py`? I would prefer not to have to enumerate the relevant roles explicitly.

**Comment by paulfantom** [comment]
Overall it looks very good! :+1:

Some nits:
- If it is possible I would use `TRAVIS_PYTHON_VERSION` instead of `LSRENV` (travis variables: https://docs.travis-ci.com/user/environment-variables/#default-environment-variables)
- `dist: xenial` is not needed as this is default in travis (source: https://docs.travis-ci.com/user/reference/overview/#linux)

----

> Would be enough to just test for the presence of {library,module_utils,*plugins}/*.py

I would say that this would be enough. However this couldn't be done in travis config. There are 3 ways of achieving this:
1) current conditional in travis config
2) bash wrapper
3) moving python environments into tox.

All of those has pros and cons. First one is nice because it won't spawn unnecessary instances but configuration becomes messy. Second one will be slower as it needs to spawn unnecessary instances (but in parallel) but gives more options for automation and auto discovery. Lastly, third option (my favorite) is to use tox and python virtual environments and specify python versions in tox. This way every dev can start the same pipline on a local machine and not only in travis.

**Comment by pcahyna** [comment]
@nkinder would this be useful for the logging role as well? I think you did molecule and travis setup for logging a bit differently and we would like to do it the same way for all the roles.

**Comment by i386x** [comment]
Hi Pawel,

Thank You for Your suggestions. After some research and experiments, I am for the finding compromise between all the 3 ways You mentioned.

Whenever `.travis.yml` starts with `language: python`, it creates python virtual environment implicitly, and in that environment it executes `tox`. As we need Python 2.6, Travis creates at least two jobs, so spawning unnecessary instances is inevitable for some roles. There is also a possibility to have just one job in which desired python versions are intalled using `sudo apt-get install -y pythonX.Y` as a system pythons, but we need to do this in `trusty`, because there is no `python2.6` package in `xenial`. Then `tox` create python virtual environments from corresponding system pythons (or skips this step if the corresponding system pythons are not installed).

Finally, I found that it is possible to set the environment variable per repository after login to https://travis-ci.org/ and going to the repository settings page. In this case, we can add something like `LSR_USE_PYTHON` environment variable, with its value set to `1`, to the role repository settings that tells that role contain some python code that should be pylinted and pytested. Then, in our `.travis.yml`, we can use `if: env(LSR_USE_PYTHON)` to tell the travis to spawn or to skip the jobs that are intended to do only tasks related to python code.

The idea to run `tox` also locally is a good catch so `envlist` is back in the game.

**Comment by i386x** [comment]
@paulfantom 
> * `dist: xenial` is not needed as this is default in travis (source: https://docs.travis-ci.com/user/reference/overview/#linux)

I am planning to add `dist: xenial` back, because of one problem that I didn't notice earlier: when Travis team bumps default Linux distro used as build environment, it may also cause the bump of system python and then it may break running the molecule for some roles (i.e. network role, where `selinux` is needed and python interpreters running in virtual env can't see C bindings if the version of python interpreter in virtual env differs from the version of system python). Currently, the molecule is run under the 3.5 Travis job, because 3.5 is a version of system python in xenial - the reason why this is hardwired is because tox-travis has no support for system python environment detection, see https://github.com/tox-dev/tox-travis/issues/128.

**Comment by richm** [comment]
I think this is mostly done, although it is different from the proposal in several places.  See the current code https://github.com/linux-system-roles/template and this PR https://github.com/linux-system-roles/template/pull/18 based on feedback from network, storage, and other roles.

What should we do with this proposal?  Update it?  Close it?

**Comment by richm** [comment]
Is this satisfied by https://github.com/linux-system-roles/tox-lsr ?
